CI: ../../gccrs/gcc/rust/util/rust-optional.h:114:7: error: '*(unsigned int*)((char*)&old_module + offsetof(Rust::Optional<unsigned int>,Rust::Optional<unsigned int>::<unnamed>))' may be used uninitialized [-Werror=maybe-uninitialized]

Hi!

I spotted this new warning (I'm building with `-Werror`) (this is with 2d806801404fc7628a3d022da92b73cd06197f49):

```
[all 2023-05-28 15:22:48] /usr/lib/gcc-snapshot/bin/g++  -fno-PIE -c  -DIN_GCC_FRONTEND -g -O2   -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE   -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Wconditionally-supported -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings -Werror -Wno-unused-parameter -fno-common  -DHAVE_CONFIG_H -I. -Irust -I../../gccrs/gcc -I../../gccrs/gcc/rust -I../../gccrs/gcc/../include  -I../../gccrs/gcc/../libcpp/include -I../../gccrs/gcc/../libcody  -I../../gccrs/gcc/../libdecnumber -I../../gccrs/gcc/../libdecnumber/dpd -I../libdecnumber -I../../gccrs/gcc/../libbacktrace   -o rust/rust-privacy-reporter.o -MT rust/rust-privacy-reporter.o -MMD -MP -MF rust/.deps/rust-privacy-reporter.TPo -g -O2   -I ../../gccrs/gcc/rust -I ../../gccrs/gcc/rust/lex -I ../../gccrs/gcc/rust/parse -I ../../gccrs/gcc/rust/ast -I ../../gccrs/gcc/rust/analysis -I ../../gccrs/gcc/rust/backend -I ../../gccrs/gcc/rust/expand -I ../../gccrs/gcc/rust/hir/tree -I ../../gccrs/gcc/rust/hir -I ../../gccrs/gcc/rust/resolve -I ../../gccrs/gcc/rust/util -I ../../gccrs/gcc/rust/typecheck -I ../../gccrs/gcc/rust/checks/lints -I ../../gccrs/gcc/rust/checks/errors -I ../../gccrs/gcc/rust/checks/errors/privacy -I ../../gccrs/gcc/rust/util -I ../../gccrs/gcc/rust/metadata -I ../../gccrs/gcc/../libgrust ../../gccrs/gcc/rust/checks/errors/privacy/rust-privacy-reporter.cc
[all 2023-05-28 15:22:59] In file included from ../../gccrs/gcc/rust/util/rust-hir-map.h:22,
[all 2023-05-28 15:22:59]                  from ../../gccrs/gcc/rust/checks/errors/privacy/rust-privacy-reporter.h:22,
[all 2023-05-28 15:22:59]                  from ../../gccrs/gcc/rust/checks/errors/privacy/rust-privacy-reporter.cc:19:
[all 2023-05-28 15:22:59] In member function 'Rust::Optional<T>& Rust::Optional<T>::operator=(const Rust::Optional<T>&) [with T = unsigned int]',
[all 2023-05-28 15:22:59]     inlined from 'virtual void Rust::Privacy::PrivacyReporter::visit(Rust::HIR::Module&)' at ../../gccrs/gcc/rust/checks/errors/privacy/rust-privacy-reporter.cc:594:20:
[all 2023-05-28 15:22:59] ../../gccrs/gcc/rust/util/rust-optional.h:114:7: error: '*(unsigned int*)((char*)&old_module + offsetof(Rust::Optional<unsigned int>,Rust::Optional<unsigned int>::<unnamed>))' may be used uninitialized [-Werror=maybe-uninitialized]
[all 2023-05-28 15:22:59]   114 |       new (&value) T (other.value);
[all 2023-05-28 15:22:59]       |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
[all 2023-05-28 15:22:59] ../../gccrs/gcc/rust/checks/errors/privacy/rust-privacy-reporter.cc: In member function 'virtual void Rust::Privacy::PrivacyReporter::visit(Rust::HIR::Module&)':
[all 2023-05-28 15:22:59] ../../gccrs/gcc/rust/checks/errors/privacy/rust-privacy-reporter.cc:587:8: note: '*(unsigned int*)((char*)&old_module + offsetof(Rust::Optional<unsigned int>,Rust::Optional<unsigned int>::<unnamed>))' was declared here
[all 2023-05-28 15:22:59]   587 |   auto old_module = current_module;
[all 2023-05-28 15:22:59]       |        ^~~~~~~~~~
[all 2023-05-28 15:23:00] cc1plus: all warnings being treated as errors
[all 2023-05-28 15:23:00] make[1]: *** [../../gccrs/gcc/rust/Make-lang.in:435: rust/rust-privacy-reporter.o] Error 1
[all 2023-05-28 15:23:00] rm m2/gm2-compiler-boot/P2Build.mod m2/gm2-compiler-boot/P0SyntaxCheck.mod m2/gm2-compiler-boot/PCBuild.mod m2/gm2-compiler-boot/PHBuild.mod m2/gm2-compiler-boot/P1Build.mod m2/gm2-compiler-boot/P3Build.mod
[all 2023-05-28 15:23:00] make[1]: Leaving directory '/var/lib/laminar/run/gccrs-alpha-linux/11/toolchain-build/gcc'
[all 2023-05-28 15:23:00] make: *** [Makefile:4667: all-gcc] Error 2
```
That's for `--target=alpha-linux` (http://toolchain.lug-owl.de/laminar/jobs/gccrs-alpha-linux/11), but I guess it's generic (ie. seen on http://toolchain.lug-owl.de/laminar/jobs/gccrs-hppa64-linux/9 and http://toolchain.lug-owl.de/laminar/jobs/gccrs-aarch64-linux/9 as well.)

--- a/gcc/rust/util/rust-optional.h
+++ b/gcc/rust/util/rust-optional.h
@@ -86,6 +86,10 @@ private:
   union
   {
     T value;
+    // prevents initialization warnings
+    // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80635#c53
+    // FIXME: remove
+    volatile char unused;
   };
 
   Optional (tag_some, const T &value) : field_is_some (true), value (value) {}

